Fiber needs mmap(2) MAP_STACK.

Add fibers for powerpc64 ELFv2, and fix fibers for 32-bit powerpc ELF.
The old code was for some other platform (AIX?), not BSD nor Linux.
This patch fixes OpenBSD but might break the other platform.

Add fibers for mips64, riscv64 and sparc64 ELF for OpenBSD only.

Index: libphobos/libdruntime/core/thread/fiber.d
--- libphobos/libdruntime/core/thread/fiber.d.orig
+++ libphobos/libdruntime/core/thread/fiber.d
@@ -125,6 +125,7 @@ private
         {
             version = AsmPPC_Posix;
             version = AsmExternal;
+            version = AlignFiberStackTo16Byte;
         }
     }
     else version (PPC64)
@@ -137,6 +138,8 @@ private
         }
         else version (Posix)
         {
+            version = AsmPPC_Posix;
+            version = AsmExternal;
             version = AlignFiberStackTo16Byte;
         }
     }
@@ -148,6 +151,15 @@ private
             version = AsmExternal;
         }
     }
+    else version (MIPS_N64)
+    {
+        version (Posix)
+        {
+            version = AsmMIPS_N64_Posix;
+            version = AsmExternal;
+            version = AlignFiberStackTo16Byte;
+        }
+    }
     else version (AArch64)
     {
         version (Posix)
@@ -172,8 +184,22 @@ private
     }
     else version (SPARC64)
     {
+        version (Posix)
+        {
+            version = AsmSPARC64_Posix;
+            version = AsmExternal;
+        }
         version = AlignFiberStackTo16Byte;
     }
+    else version (RISCV64)
+    {
+        version (Posix)
+        {
+            version = AsmRISCV64_Posix;
+            version = AsmExternal;
+            version = AlignFiberStackTo16Byte;
+        }
+    }
 
     version (Posix)
     {
@@ -236,6 +262,12 @@ private
       extern (C) void fiber_switchContext( void** oldp, void* newp ) nothrow @nogc;
       version (AArch64)
           extern (C) void fiber_trampoline() nothrow;
+      version (MIPS_N64)
+          extern (C) void fiber_trampoline() nothrow;
+      version (SPARC64)
+          extern (C) void fiber_trampoline() nothrow;
+      version (RISCV64)
+          extern (C) void fiber_trampoline() nothrow;
   }
   else
     extern (C) void fiber_switchContext( void** oldp, void* newp ) nothrow @nogc
@@ -1033,10 +1065,13 @@ class Fiber (private)
                 // Allocate more for the memory guard
                 sz += guardPageSize;
 
+                int mmap_flags = MAP_PRIVATE | MAP_ANON;
+                version (OpenBSD)
+                    mmap_flags |= MAP_STACK;
                 m_pmem = mmap( null,
                                sz,
                                PROT_READ | PROT_WRITE,
-                               MAP_PRIVATE | MAP_ANON,
+                               mmap_flags,
                                -1,
                                0 );
                 if ( m_pmem == MAP_FAILED )
@@ -1338,27 +1373,22 @@ class Fiber (private)
         }
         else version (AsmPPC_Posix)
         {
-            version (StackGrowsDown)
-            {
-                pstack -= int.sizeof * 5;
-            }
-            else
-            {
-                pstack += int.sizeof * 5;
-            }
+            version (StackGrowsDown) {} else static assert (false);
 
-            push( cast(size_t) &fiber_entryPoint );     // link register
-            push( 0x00000000 );                         // control register
-            push( 0x00000000 );                         // old stack pointer
-
-            // GPR values
-            version (StackGrowsDown)
+            version (PPC64)
             {
-                pstack -= int.sizeof * 20;
+                pstack -= size_t.sizeof * 5;
+                push( cast(size_t) &fiber_entryPoint);  // link register
+                push( 0x00000000_00000000 );            // condition register
+                push( 0x00000000_00000000 );            // old stack pointer
+                pstack -= size_t.sizeof * 18;           // GPRs
             }
             else
             {
-                pstack += int.sizeof * 20;
+                pstack -= size_t.sizeof * 6;
+                push( cast(size_t) &fiber_entryPoint);  // link register
+                push( 0x00000000 );                     // old stack pointer
+                pstack -= size_t.sizeof * 20;           // GPRs and CR
             }
 
             assert( (cast(size_t) pstack & 0x0f) == 0 );
@@ -1426,6 +1456,23 @@ class Fiber (private)
             pstack -= ABOVE;
             *cast(size_t*)(pstack - SZ_RA) = cast(size_t)&fiber_entryPoint;
         }
+        else version (AsmMIPS_N64_Posix)
+        {
+            version (StackGrowsDown) {}
+            else static assert(false, "Only full descending stacks supported on MIPS N64");
+
+            enum FRAME_SIZE = 160;
+            enum TSTACK_OFFSET = 80;
+            enum RA_OFFSET = 72;
+            enum S0_OFFSET = TSTACK_OFFSET;
+
+            (cast(ubyte*)pstack - FRAME_SIZE)[0 .. FRAME_SIZE] = 0;
+            pstack -= FRAME_SIZE;
+            *cast(size_t*)(pstack + RA_OFFSET) = cast(size_t) &fiber_trampoline;
+            // The trampoline uses saved s0 to establish t9 for PIC entry.
+            *cast(size_t*)(pstack + S0_OFFSET) = cast(size_t) &fiber_entryPoint;
+            pstack += TSTACK_OFFSET;
+        }
         else version (AsmAArch64_Posix)
         {
             // Like others, FP registers and return address (lr) are kept
@@ -1448,6 +1495,46 @@ class Fiber (private)
             pstack -= size_t.sizeof * 11;    // skip past x19-x29
             push(cast(size_t) &fiber_trampoline); // see threadasm.S for docs
             pstack += size_t.sizeof;         // adjust sp (newp) above lr
+        }
+        else version (AsmSPARC64_Posix)
+        {
+            // FP regs and return address kept below tstack to hide from GC.
+            // See switchcontext.S for frame layout. tstack points to %l0.
+            //   [0-127]   - reserved (register window save area)
+            //   [RSA+0]   - %o7 (return address)
+            //   [RSA+8]   - %f32-%f62
+            //   [RSA+136] - %l0-%l7  <-- tstack
+            //   [RSA+200] - %i0-%i7
+
+            version (StackGrowsDown) {}
+            else
+                static assert(false, "Only full descending stacks supported on SPARC64");
+
+            enum RSA = 128;
+            enum FRAME_SIZE = 400;
+            enum TSTACK_OFFSET = RSA + 136;
+
+            (cast(ubyte*)pstack - (FRAME_SIZE + RSA))[0 .. FRAME_SIZE + RSA] = 0;
+            pstack -= FRAME_SIZE + RSA;
+            *cast(size_t*)(pstack + RSA) = cast(size_t) &fiber_trampoline - 8;
+            pstack += TSTACK_OFFSET;
+        }
+        else version (AsmRISCV64_Posix)
+        {
+            // FP regs and return address are kept below tstack to hide from GC.
+            // See switchcontext.S for frame layout. tstack points to s0.
+            version (StackGrowsDown) {}
+            else
+                static assert(false, "Only full descending stacks supported on RISCV64");
+
+            enum FRAME_SIZE = 208;
+            enum TSTACK_OFFSET = 112;
+            enum RA_OFFSET = 96;
+
+            (cast(ubyte*)pstack - FRAME_SIZE)[0 .. FRAME_SIZE] = 0;
+            pstack -= FRAME_SIZE;
+            *cast(size_t*)(pstack + RA_OFFSET) = cast(size_t) &fiber_trampoline;
+            pstack += TSTACK_OFFSET;
         }
         else version (AsmARM_Posix)
         {
